{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5ccbabfc-cabd-429a-bc37-609a0b6bbed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7c29464-cb8a-483a-9121-2638bf35fc0d",
   "metadata": {},
   "source": [
    "Metadata file has multiple columns. The data comes from CMV status analysis (though we don't need it here). We need to select those items which are homozygous by A\\*02 HLA. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ed1d6310-2af8-402e-bc59-b5ced6927ac2",
   "metadata": {},
   "outputs": [],
   "source": [
    "desc_hip = pd.read_csv('/projects/fmba_covid/1_data_links/hip_full_prep.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0be02860-cf49-42fe-9002-e29dc975f802",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>age</th>\n",
       "      <th>race</th>\n",
       "      <th>sex</th>\n",
       "      <th>cmv</th>\n",
       "      <th>hla</th>\n",
       "      <th>..filter..</th>\n",
       "      <th>dataset</th>\n",
       "      <th>covid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HIP00110.txt</td>\n",
       "      <td>HIP00110</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-A*03,HLA-A*24,HLA-B*07</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>HIP</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIP00169.txt</td>\n",
       "      <td>HIP00169</td>\n",
       "      <td>41.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>male</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-A*01,HLA-A*02,HLA-B*27,HLA-B*37</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>HIP</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HIP00594.txt</td>\n",
       "      <td>HIP00594</td>\n",
       "      <td>21.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>+</td>\n",
       "      <td>HLA-A*02,HLA-A*32,HLA-B*07,HLA-B*61</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>HIP</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HIP00602.txt</td>\n",
       "      <td>HIP00602</td>\n",
       "      <td>45.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-A*02,HLA-A*11,HLA-B*35,HLA-B*51</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>HIP</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP00614.txt</td>\n",
       "      <td>HIP00614</td>\n",
       "      <td>27.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-A*03,HLA-A*29,HLA-B*07</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>HIP</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>Keck0116_MC1.txt</td>\n",
       "      <td>Keck0116_MC1</td>\n",
       "      <td>19.0</td>\n",
       "      <td>asian or pacific islander,non-hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>KECK</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>782</th>\n",
       "      <td>Keck0117_MC1.txt</td>\n",
       "      <td>Keck0117_MC1</td>\n",
       "      <td>34.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>KECK</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>783</th>\n",
       "      <td>Keck0118_MC1.txt</td>\n",
       "      <td>Keck0118_MC1</td>\n",
       "      <td>28.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>KECK</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <td>Keck0119_MC1.txt</td>\n",
       "      <td>Keck0119_MC1</td>\n",
       "      <td>24.0</td>\n",
       "      <td>asian or pacific islander,non-hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>KECK</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>785</th>\n",
       "      <td>Keck0120_MC1.txt</td>\n",
       "      <td>Keck0120_MC1</td>\n",
       "      <td>26.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "      <td>KECK</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>786 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            file_name     sample_id   age  \\\n",
       "0        HIP00110.txt      HIP00110  55.0   \n",
       "1        HIP00169.txt      HIP00169  41.0   \n",
       "2        HIP00594.txt      HIP00594  21.0   \n",
       "3        HIP00602.txt      HIP00602  45.0   \n",
       "4        HIP00614.txt      HIP00614  27.0   \n",
       "..                ...           ...   ...   \n",
       "781  Keck0116_MC1.txt  Keck0116_MC1  19.0   \n",
       "782  Keck0117_MC1.txt  Keck0117_MC1  34.0   \n",
       "783  Keck0118_MC1.txt  Keck0118_MC1  28.0   \n",
       "784  Keck0119_MC1.txt  Keck0119_MC1  24.0   \n",
       "785  Keck0120_MC1.txt  Keck0120_MC1  26.0   \n",
       "\n",
       "                                                 race     sex cmv  \\\n",
       "0                                                 NaN    male   -   \n",
       "1                                                 NaN    male   -   \n",
       "2                    caucasian,non-hispanic or latino    male   +   \n",
       "3                                                 NaN  female   -   \n",
       "4                    caucasian,non-hispanic or latino    male   -   \n",
       "..                                                ...     ...  ..   \n",
       "781  asian or pacific islander,non-hispanic or latino  female   -   \n",
       "782                  caucasian,non-hispanic or latino    male   +   \n",
       "783                  caucasian,non-hispanic or latino    male   +   \n",
       "784  asian or pacific islander,non-hispanic or latino  female   +   \n",
       "785                  caucasian,non-hispanic or latino  female   -   \n",
       "\n",
       "                                     hla       ..filter.. dataset    covid  \n",
       "0             HLA-A*03,HLA-A*24,HLA-B*07  corr:2:0.05:all     HIP  healthy  \n",
       "1    HLA-A*01,HLA-A*02,HLA-B*27,HLA-B*37  corr:2:0.05:all     HIP  healthy  \n",
       "2    HLA-A*02,HLA-A*32,HLA-B*07,HLA-B*61  corr:2:0.05:all     HIP  healthy  \n",
       "3    HLA-A*02,HLA-A*11,HLA-B*35,HLA-B*51  corr:2:0.05:all     HIP  healthy  \n",
       "4             HLA-A*03,HLA-A*29,HLA-B*07  corr:2:0.05:all     HIP  healthy  \n",
       "..                                   ...              ...     ...      ...  \n",
       "781                                  NaN  corr:2:0.05:all    KECK  healthy  \n",
       "782                                  NaN  corr:2:0.05:all    KECK  healthy  \n",
       "783                                  NaN  corr:2:0.05:all    KECK  healthy  \n",
       "784                                  NaN  corr:2:0.05:all    KECK  healthy  \n",
       "785                                  NaN  corr:2:0.05:all    KECK  healthy  \n",
       "\n",
       "[786 rows x 10 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "desc_hip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bcaeeb38-d591-4071-96cf-086c297a2a8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "desc_hip.file_name = 'hip_full/' + desc_hip.file_name.apply(lambda x: x.split('/')[-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b68fc46-bb62-487f-a80f-5ae3bd866880",
   "metadata": {},
   "source": [
    "Here I select only those samples which are homozygous by HLA-A\\*02."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "79b58a9d-957b-45a7-86df-d935701b20f8",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'float' object has no attribute 'split'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Input \u001b[0;32mIn [5]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m desc_hip[\u001b[43mdesc_hip\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhla\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mlambda\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m,\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43my\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msplit\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m,\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mA*\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHLA-A*02\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[0;32m~/.conda-envs/bio/lib/python3.9/site-packages/pandas/core/series.py:4433\u001b[0m, in \u001b[0;36mSeries.apply\u001b[0;34m(self, func, convert_dtype, args, **kwargs)\u001b[0m\n\u001b[1;32m   4323\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mapply\u001b[39m(\n\u001b[1;32m   4324\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   4325\u001b[0m     func: AggFuncType,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4328\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   4329\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m Series:\n\u001b[1;32m   4330\u001b[0m     \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   4331\u001b[0m \u001b[38;5;124;03m    Invoke function on values of Series.\u001b[39;00m\n\u001b[1;32m   4332\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   4431\u001b[0m \u001b[38;5;124;03m    dtype: float64\u001b[39;00m\n\u001b[1;32m   4432\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 4433\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mSeriesApply\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconvert_dtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda-envs/bio/lib/python3.9/site-packages/pandas/core/apply.py:1082\u001b[0m, in \u001b[0;36mSeriesApply.apply\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1078\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mf, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m   1079\u001b[0m     \u001b[38;5;66;03m# if we are a string, try to dispatch\u001b[39;00m\n\u001b[1;32m   1080\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mapply_str()\n\u001b[0;32m-> 1082\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.conda-envs/bio/lib/python3.9/site-packages/pandas/core/apply.py:1137\u001b[0m, in \u001b[0;36mSeriesApply.apply_standard\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1131\u001b[0m         values \u001b[38;5;241m=\u001b[39m obj\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mobject\u001b[39m)\u001b[38;5;241m.\u001b[39m_values\n\u001b[1;32m   1132\u001b[0m         \u001b[38;5;66;03m# error: Argument 2 to \"map_infer\" has incompatible type\u001b[39;00m\n\u001b[1;32m   1133\u001b[0m         \u001b[38;5;66;03m# \"Union[Callable[..., Any], str, List[Union[Callable[..., Any], str]],\u001b[39;00m\n\u001b[1;32m   1134\u001b[0m         \u001b[38;5;66;03m# Dict[Hashable, Union[Union[Callable[..., Any], str],\u001b[39;00m\n\u001b[1;32m   1135\u001b[0m         \u001b[38;5;66;03m# List[Union[Callable[..., Any], str]]]]]\"; expected\u001b[39;00m\n\u001b[1;32m   1136\u001b[0m         \u001b[38;5;66;03m# \"Callable[[Any], Any]\"\u001b[39;00m\n\u001b[0;32m-> 1137\u001b[0m         mapped \u001b[38;5;241m=\u001b[39m \u001b[43mlib\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap_infer\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1138\u001b[0m \u001b[43m            \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1139\u001b[0m \u001b[43m            \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[arg-type]\u001b[39;49;00m\n\u001b[1;32m   1140\u001b[0m \u001b[43m            \u001b[49m\u001b[43mconvert\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconvert_dtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1141\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1143\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(mapped) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(mapped[\u001b[38;5;241m0\u001b[39m], ABCSeries):\n\u001b[1;32m   1144\u001b[0m     \u001b[38;5;66;03m# GH#43986 Need to do list(mapped) in order to get treated as nested\u001b[39;00m\n\u001b[1;32m   1145\u001b[0m     \u001b[38;5;66;03m#  See also GH#25959 regarding EA support\u001b[39;00m\n\u001b[1;32m   1146\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m obj\u001b[38;5;241m.\u001b[39m_constructor_expanddim(\u001b[38;5;28mlist\u001b[39m(mapped), index\u001b[38;5;241m=\u001b[39mobj\u001b[38;5;241m.\u001b[39mindex)\n",
      "File \u001b[0;32m~/.conda-envs/bio/lib/python3.9/site-packages/pandas/_libs/lib.pyx:2870\u001b[0m, in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "Input \u001b[0;32mIn [5]\u001b[0m, in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[0;32m----> 1\u001b[0m desc_hip[desc_hip\u001b[38;5;241m.\u001b[39mhla\u001b[38;5;241m.\u001b[39mapply(\u001b[38;5;28;01mlambda\u001b[39;00m x: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mjoin([y \u001b[38;5;28;01mfor\u001b[39;00m y \u001b[38;5;129;01min\u001b[39;00m \u001b[43mx\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msplit\u001b[49m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mA*\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m y])) \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHLA-A*02\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'float' object has no attribute 'split'"
     ]
    }
   ],
   "source": [
    "desc_hip[desc_hip.hla.apply(lambda x: ','.join([y for y in x.split(',') if 'A*' in y])) == 'HLA-A*02']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "842bc923-f430-4170-b82f-e517772471c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_data = desc_hip[desc_hip.hla.apply(lambda x: ','.join([y for y in x.split(',') if 'A*' in y])) == 'HLA-A*02'].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "829a6ecf-f75a-4d0d-ac37-174e631b2bb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>age</th>\n",
       "      <th>race</th>\n",
       "      <th>sex</th>\n",
       "      <th>cmv</th>\n",
       "      <th>hla</th>\n",
       "      <th>..filter..</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>hip_full/HIP00813.txt</td>\n",
       "      <td>HIP00813</td>\n",
       "      <td>32.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>+</td>\n",
       "      <td>HLA-A*02,HLA-B*51</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>hip_full/HIP02112.txt</td>\n",
       "      <td>HIP02112</td>\n",
       "      <td>6.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-A*02,HLA-B*07,HLA-B*15</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>hip_full/HIP03484.txt</td>\n",
       "      <td>HIP03484</td>\n",
       "      <td>46.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-A*02,HLA-B*07,HLA-B*58</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>hip_full/HIP03495.txt</td>\n",
       "      <td>HIP03495</td>\n",
       "      <td>62.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>+</td>\n",
       "      <td>HLA-A*02,HLA-B*62</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>hip_full/HIP03618.txt</td>\n",
       "      <td>HIP03618</td>\n",
       "      <td>62.0</td>\n",
       "      <td>caucasian,non-hispanic or latino</td>\n",
       "      <td>male</td>\n",
       "      <td>+</td>\n",
       "      <td>HLA-A*02,HLA-B*35,HLA-B*44</td>\n",
       "      <td>corr:2:0.05:all</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 file_name sample_id   age                              race  \\\n",
       "19   hip_full/HIP00813.txt  HIP00813  32.0  caucasian,non-hispanic or latino   \n",
       "91   hip_full/HIP02112.txt  HIP02112   6.0  caucasian,non-hispanic or latino   \n",
       "131  hip_full/HIP03484.txt  HIP03484  46.0  caucasian,non-hispanic or latino   \n",
       "133  hip_full/HIP03495.txt  HIP03495  62.0  caucasian,non-hispanic or latino   \n",
       "140  hip_full/HIP03618.txt  HIP03618  62.0  caucasian,non-hispanic or latino   \n",
       "\n",
       "        sex cmv                         hla       ..filter..  \n",
       "19     male   +           HLA-A*02,HLA-B*51  corr:2:0.05:all  \n",
       "91   female   -  HLA-A*02,HLA-B*07,HLA-B*15  corr:2:0.05:all  \n",
       "131    male   -  HLA-A*02,HLA-B*07,HLA-B*58  corr:2:0.05:all  \n",
       "133  female   +           HLA-A*02,HLA-B*62  corr:2:0.05:all  \n",
       "140    male   +  HLA-A*02,HLA-B*35,HLA-B*44  corr:2:0.05:all  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "da81b1d1-224c-42af-b60b-4bf1b4af33dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "for run in selected_data.file_name:\n",
    "    shutil.copyfile('/projects/fmba_covid/1_data_links/' + run, run)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0da918c-b76f-46f3-b9ab-fa3e0b0735e0",
   "metadata": {},
   "source": [
    "Reading the raw rep-seq data. We can find the info on V/D/J segments which were used to form this clonotype, its frequency in sample, count of times it was assembled using spesific nucleotide sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ac78900a-7788-4b1f-a652-2a6f611064c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>freq</th>\n",
       "      <th>cdr3nt</th>\n",
       "      <th>cdr3aa</th>\n",
       "      <th>v</th>\n",
       "      <th>d</th>\n",
       "      <th>j</th>\n",
       "      <th>VEnd</th>\n",
       "      <th>DStart</th>\n",
       "      <th>DEnd</th>\n",
       "      <th>JStart</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>42662</td>\n",
       "      <td>0.025387</td>\n",
       "      <td>TGCGCCAGCAGCCCCTGGAGGGGAGACACTGAAGCTTTCTTT</td>\n",
       "      <td>CASSPWRGDTEAFF</td>\n",
       "      <td>TRBV4-1,TRBV4-3</td>\n",
       "      <td>TRBD2</td>\n",
       "      <td>TRBJ1-1</td>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22539</td>\n",
       "      <td>0.013412</td>\n",
       "      <td>TGTGCCAGCAGTCAAACGGGACACTGAGAACACTGAAGCTTTCTTT</td>\n",
       "      <td>CASSQTGH_ENTEAFF</td>\n",
       "      <td>TRBV6-4</td>\n",
       "      <td>TRBD1,TRBD2</td>\n",
       "      <td>TRBJ1-1</td>\n",
       "      <td>12</td>\n",
       "      <td>16</td>\n",
       "      <td>23</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13571</td>\n",
       "      <td>0.008076</td>\n",
       "      <td>TGTGCCAGCTCTCCCCGGGACAGGAGACAAGAGACCCAGTACTTC</td>\n",
       "      <td>CASSPRDRRQETQYF</td>\n",
       "      <td>TRBV6-5,TRBV6-1,TRBV6-6</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>TRBJ2-5</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>24</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13229</td>\n",
       "      <td>0.007872</td>\n",
       "      <td>TGTGCCAGCAGTTTAGGGCTCGGGAGCTCCTATAATTCACCCCTCC...</td>\n",
       "      <td>CASSLGLGSSYNSPLHF</td>\n",
       "      <td>TRBV28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-6</td>\n",
       "      <td>15</td>\n",
       "      <td>-31</td>\n",
       "      <td>-31</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7341</td>\n",
       "      <td>0.004368</td>\n",
       "      <td>TGTGCCAGCAGTGAAGCGGGGGGCACAGATACGCAGTATTTT</td>\n",
       "      <td>CASSEAGGTDTQYF</td>\n",
       "      <td>TRBV6-4</td>\n",
       "      <td>TRBD2</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114118</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TGTGCCACCAGCAGAGCGAGCAGGGCCTCCAGCGGGGCCAACGTCC...</td>\n",
       "      <td>CATSRASRASSGANVLTF</td>\n",
       "      <td>TRBV15</td>\n",
       "      <td>TRBD2</td>\n",
       "      <td>TRBJ2-6</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "      <td>33</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114119</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TGTGCCACCAGCAGAGTCCCCGTATCGACAGGGGTTAAGAATGAGC...</td>\n",
       "      <td>CATSRVPVSTGVKNEQFF</td>\n",
       "      <td>TRBV15</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>16</td>\n",
       "      <td>26</td>\n",
       "      <td>34</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114120</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TGTGCCACCAGCAGCCTATCGGGACAGGAAGGTTCCTACAATGAGC...</td>\n",
       "      <td>CATSSLSGQEGSYNEQFF</td>\n",
       "      <td>TRBV15</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>14</td>\n",
       "      <td>19</td>\n",
       "      <td>28</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114121</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TGTGCCACCAGCAGTCTTATGCGACAGGGACATTACAATCAGCCCC...</td>\n",
       "      <td>CATSSLMRQGHYNQPQHF</td>\n",
       "      <td>TRBV15</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>TRBJ1-5</td>\n",
       "      <td>14</td>\n",
       "      <td>22</td>\n",
       "      <td>29</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114122</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>TGTATCTCTGTGTCAGAAGCTTCCACTCGGGCAACGGGACAGCACA...</td>\n",
       "      <td>CISVSEASTRA_GTAQANYGYTF</td>\n",
       "      <td>TRBV7-5</td>\n",
       "      <td>TRBD1</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "      <td>22</td>\n",
       "      <td>34</td>\n",
       "      <td>42</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114123 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        count      freq                                             cdr3nt  \\\n",
       "0       42662  0.025387         TGCGCCAGCAGCCCCTGGAGGGGAGACACTGAAGCTTTCTTT   \n",
       "1       22539  0.013412     TGTGCCAGCAGTCAAACGGGACACTGAGAACACTGAAGCTTTCTTT   \n",
       "2       13571  0.008076      TGTGCCAGCTCTCCCCGGGACAGGAGACAAGAGACCCAGTACTTC   \n",
       "3       13229  0.007872  TGTGCCAGCAGTTTAGGGCTCGGGAGCTCCTATAATTCACCCCTCC...   \n",
       "4        7341  0.004368         TGTGCCAGCAGTGAAGCGGGGGGCACAGATACGCAGTATTTT   \n",
       "...       ...       ...                                                ...   \n",
       "114118      2  0.000001  TGTGCCACCAGCAGAGCGAGCAGGGCCTCCAGCGGGGCCAACGTCC...   \n",
       "114119      2  0.000001  TGTGCCACCAGCAGAGTCCCCGTATCGACAGGGGTTAAGAATGAGC...   \n",
       "114120      2  0.000001  TGTGCCACCAGCAGCCTATCGGGACAGGAAGGTTCCTACAATGAGC...   \n",
       "114121      2  0.000001  TGTGCCACCAGCAGTCTTATGCGACAGGGACATTACAATCAGCCCC...   \n",
       "114122      2  0.000001  TGTATCTCTGTGTCAGAAGCTTCCACTCGGGCAACGGGACAGCACA...   \n",
       "\n",
       "                         cdr3aa                        v            d  \\\n",
       "0                CASSPWRGDTEAFF          TRBV4-1,TRBV4-3        TRBD2   \n",
       "1              CASSQTGH_ENTEAFF                  TRBV6-4  TRBD1,TRBD2   \n",
       "2               CASSPRDRRQETQYF  TRBV6-5,TRBV6-1,TRBV6-6        TRBD1   \n",
       "3             CASSLGLGSSYNSPLHF                   TRBV28          NaN   \n",
       "4                CASSEAGGTDTQYF                  TRBV6-4        TRBD2   \n",
       "...                         ...                      ...          ...   \n",
       "114118       CATSRASRASSGANVLTF                   TRBV15        TRBD2   \n",
       "114119       CATSRVPVSTGVKNEQFF                   TRBV15        TRBD1   \n",
       "114120       CATSSLSGQEGSYNEQFF                   TRBV15        TRBD1   \n",
       "114121       CATSSLMRQGHYNQPQHF                   TRBV15        TRBD1   \n",
       "114122  CISVSEASTRA_GTAQANYGYTF                  TRBV7-5        TRBD1   \n",
       "\n",
       "              j  VEnd  DStart  DEnd  JStart  \n",
       "0       TRBJ1-1    13      16    22      25  \n",
       "1       TRBJ1-1    12      16    23      27  \n",
       "2       TRBJ2-5     9      13    24      27  \n",
       "3       TRBJ1-6    15     -31   -31      22  \n",
       "4       TRBJ2-3    14      14    20      22  \n",
       "...         ...   ...     ...   ...     ...  \n",
       "114118  TRBJ2-6    16      21    33      33  \n",
       "114119  TRBJ2-1    16      26    34      39  \n",
       "114120  TRBJ2-1    14      19    28      33  \n",
       "114121  TRBJ1-5    14      22    29      35  \n",
       "114122  TRBJ1-2    22      34    42      46  \n",
       "\n",
       "[114123 rows x 11 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('/projects/fmba_covid/1_data_links/hip_full/HIP00813.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0ff7257-f55b-4e42-a0d5-3ab7170ac793",
   "metadata": {},
   "source": [
    "Reading theough all the five raw files and concatenating them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a507df44-9fbd-44d1-b310-4fafedd37db5",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_to_concat = []\n",
    "for run in selected_data.file_name:\n",
    "    datasets_to_concat.append(pd.read_csv('/projects/fmba_covid/1_data_links/' + run, sep='\\t'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ff9ef4b9-2a0e-4b7a-8100-0359cee91290",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_data_concat = pd.concat(datasets_to_concat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "435f6747-5e07-4154-8416-67565b2045a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_data_concat = raw_data_concat[raw_data_concat.cdr3aa.str.isalpha()][['cdr3aa', 'count']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22017c36-c102-4a62-a6e9-ac4c36e9a280",
   "metadata": {},
   "source": [
    "We want to have a look at the naive clonotypes. It means that they would be very rare (assembpled only in one sample, only once). To choose such clonotypes let's group them:\n",
    "* by how many samples have this clonotype\n",
    "* by how many cells we sequenced which correspond to this cdr3aa (amino acid sequence of CDR3)\n",
    "\n",
    "As the result we would take the intersection of these two sets of clonotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "75bdbcd7-e65b-4737-a443-2d921e53bae6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3aa</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AASSEATWGGPYEQYF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604797</th>\n",
       "      <td>CASSSPGRSNYGYTF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604798</th>\n",
       "      <td>CASSSPGRSSGNTIYF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604799</th>\n",
       "      <td>CASSSPGRSSSYEQYF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604801</th>\n",
       "      <td>CASSSPGRTDTEAFF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384690</th>\n",
       "      <td>CASSLSSYEQYF</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315117</th>\n",
       "      <td>CASSLGETQYF</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338319</th>\n",
       "      <td>CASSLGYEQYF</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320231</th>\n",
       "      <td>CASSLGGTEAFF</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318469</th>\n",
       "      <td>CASSLGGNTEAFF</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>914785 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  cdr3aa  count\n",
       "0       AASSEATWGGPYEQYF      1\n",
       "604797   CASSSPGRSNYGYTF      1\n",
       "604798  CASSSPGRSSGNTIYF      1\n",
       "604799  CASSSPGRSSSYEQYF      1\n",
       "604801   CASSSPGRTDTEAFF      1\n",
       "...                  ...    ...\n",
       "384690      CASSLSSYEQYF     48\n",
       "315117       CASSLGETQYF     50\n",
       "338319       CASSLGYEQYF     51\n",
       "320231      CASSLGGTEAFF     51\n",
       "318469     CASSLGGNTEAFF     61\n",
       "\n",
       "[914785 rows x 2 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_data_per_file = raw_data_concat.groupby(by=['cdr3aa'], as_index=False).count()\n",
    "grouped_data_per_file.sort_values(by='count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "18f63a07-9124-487a-a749-139028a53009",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3aa</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AASSEATWGGPYEQYF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AASSLNLPPANSPLHF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AASSQVAEGGFYGYTF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>APGYSSNQPQHF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AQDTQYF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914780</th>\n",
       "      <td>YSVPGVSLAGYGYTF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914781</th>\n",
       "      <td>YTGVFQAGPSAPAEQYF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914782</th>\n",
       "      <td>YTSINGTSGNQDEQFF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914783</th>\n",
       "      <td>YVYEQYF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914784</th>\n",
       "      <td>YWDNKGNTEAFF</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>852420 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   cdr3aa  count\n",
       "0        AASSEATWGGPYEQYF      1\n",
       "1        AASSLNLPPANSPLHF      1\n",
       "2        AASSQVAEGGFYGYTF      1\n",
       "3            APGYSSNQPQHF      1\n",
       "4                 AQDTQYF      1\n",
       "...                   ...    ...\n",
       "914780    YSVPGVSLAGYGYTF      1\n",
       "914781  YTGVFQAGPSAPAEQYF      1\n",
       "914782   YTSINGTSGNQDEQFF      1\n",
       "914783            YVYEQYF      1\n",
       "914784       YWDNKGNTEAFF      1\n",
       "\n",
       "[852420 rows x 2 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_clones_per_file = grouped_data_per_file[grouped_data_per_file['count'] == 1]\n",
    "unique_clones_per_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "bdce70a2-27e7-49ac-abc9-d496703d34e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3aa</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>117737</th>\n",
       "      <td>CASRRQGAGYEQYF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>742957</th>\n",
       "      <td>CATNREMVTNTGELFF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>887844</th>\n",
       "      <td>CSSSLVRAGDYEQYF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84273</th>\n",
       "      <td>CASRDGQGAWETQYF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84271</th>\n",
       "      <td>CASRDGQGATNEKLFF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328042</th>\n",
       "      <td>CASSLGQGRPQHF</td>\n",
       "      <td>59893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666320</th>\n",
       "      <td>CASSVVPGQNYEQYF</td>\n",
       "      <td>64161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357506</th>\n",
       "      <td>CASSLPGQTTYEQYF</td>\n",
       "      <td>84247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>604184</th>\n",
       "      <td>CASSSPGLIGHEQYF</td>\n",
       "      <td>97769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150013</th>\n",
       "      <td>CASSDGKTSWGTEAFF</td>\n",
       "      <td>372266</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>914785 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  cdr3aa   count\n",
       "117737    CASRRQGAGYEQYF       2\n",
       "742957  CATNREMVTNTGELFF       2\n",
       "887844   CSSSLVRAGDYEQYF       2\n",
       "84273    CASRDGQGAWETQYF       2\n",
       "84271   CASRDGQGATNEKLFF       2\n",
       "...                  ...     ...\n",
       "328042     CASSLGQGRPQHF   59893\n",
       "666320   CASSVVPGQNYEQYF   64161\n",
       "357506   CASSLPGQTTYEQYF   84247\n",
       "604184   CASSSPGLIGHEQYF   97769\n",
       "150013  CASSDGKTSWGTEAFF  372266\n",
       "\n",
       "[914785 rows x 2 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_data_per_sum = raw_data_concat.groupby(by=['cdr3aa'], as_index=False).sum()\n",
    "grouped_data_per_sum.sort_values(by='count')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4699375d-ced4-477d-9113-6fc8dcd810b2",
   "metadata": {},
   "source": [
    "We can see that there are no clonotypes which were assembpled only once, but there are many which were assempled twice (mb some effect of sequencing features). Let's assume clonotypes with count equal to 2 as naive."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f8c4a9c2-a55f-4697-98a2-9a465152e013",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3aa</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CAAAGGETQYF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CAAAGGPGNEQFF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CAAAGLAKNIQYF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CAAAKSRDTQYF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>CAAARDFSYEQYF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914769</th>\n",
       "      <td>YSASGQGAYEQFF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914770</th>\n",
       "      <td>YSASPDGLPQPQHF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914771</th>\n",
       "      <td>YSATSSDTQYF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914773</th>\n",
       "      <td>YSGQAWGEQFF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>914776</th>\n",
       "      <td>YSSSLDQSKRNNSPLHF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74055 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   cdr3aa  count\n",
       "26            CAAAGGETQYF      2\n",
       "29          CAAAGGPGNEQFF      2\n",
       "31          CAAAGLAKNIQYF      2\n",
       "47           CAAAKSRDTQYF      2\n",
       "64          CAAARDFSYEQYF      2\n",
       "...                   ...    ...\n",
       "914769      YSASGQGAYEQFF      2\n",
       "914770     YSASPDGLPQPQHF      2\n",
       "914771        YSATSSDTQYF      2\n",
       "914773        YSGQAWGEQFF      2\n",
       "914776  YSSSLDQSKRNNSPLHF      2\n",
       "\n",
       "[74055 rows x 2 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_clones_per_sum = grouped_data_per_sum[grouped_data_per_sum['count'] == 2]\n",
    "unique_clones_per_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "fd7c8bae-267a-4152-9d98-7f94833f8902",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "74055"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(unique_clones_per_file.cdr3aa).intersection(set(unique_clones_per_sum.cdr3aa)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d06cbd07-ef43-4b11-9fa2-49ba9fdf93f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_clones_per_sum[['cdr3aa']].to_csv('naive_clonotypes.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "17cc4e71-c18a-44e9-aefb-0a67418cd83f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import binom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "25dcb576-1407-4167-af08-1290062287e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0016349373999999981"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binom.cdf(5.5, 10, 0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41cd4d9f-2fbe-4f78-b277-db915eae2e2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in range()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio",
   "language": "python",
   "name": "bio"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
